[toc]

# 1 什么是协程

学习协程遇到的第一个问题是：什么是协程？一个简短的回答是：协程是一个函数，它可以暂停以及恢复执行。按照我们对普通函数的理解，函数暂停意味着线程停止运行了（就像命中了断点一样），那协程的不同之处在哪里呢？区别在于，普通函数是线程相关的，函数的状态跟线程紧密关联；而协程是线程无关的，它的状态与任何线程都没有关系。

协程不一样，协程的状态是保存在堆内存上的。假设 Bar() 是一个协程，那么调用它的过程如下所示：
<img src="https://zplutor.github.io/2022/03/25/cpp-coroutine-beginner/2.png">

首先， Bar() 的状态所需的内存会在堆上分配，独立于线程栈而存在。传递给它的参数都会复制到这个状态中，而局部变量会直接在这个状态中创建。调用 Bar() 的时候，由于本质上还是一个函数调用，所以栈顶指针也会往下移动，在栈上给执行 Bar() 所需的状态分配空间，其中会有一个引用指向在堆上的状态，这样一来， Bar() 就可以像一个普通函数那样执行了，线程也可以访问到位于堆上的状态。


如果协程需要暂停，那么当前执行到的代码位置会记录到堆的状态中。然后栈上的执行时状态被销毁，栈顶指针移动以回收空间，就像普通函数结束时那样。在下一次恢复执行时，堆状态中记录的暂停位置会读取出来，从这个位置接着执行。这样就实现了一个可暂停和恢复执行的函数。

由此可见，当协程执行的时候，它跟普通函数一样，也是需要依赖线程栈；但是，一旦它暂停了，它的状态就会独立保存在堆中，此时它跟任何线程都没有关系，调用它的线程可以继续去做其它事情而不会停止。在下一次恢复执行时，协程可以由上次执行的线程来执行，也可以由另外一个完全不同的线程来执行。所以说协程是线程无关的。


# 2 实现一个协程
在C++中，只要在函数体内出现了 co_await 、co_return 和 co_yield 这三个操作符中的其中一个，这个函数就成为了协程。

## co_await
co_await 的作用是让协程暂停下来，等待某个操作完成之后再恢复执行。在使用co_await的时候还需要知道什么时候执行完毕、怎么获取执行结果，所以co_await还需要实现以下：

* await_ready()，返回类型是 bool。协程在执行 co_await 的时候，会先调用 await_ready() 来询问“操作是否已完成”，如果函数返回了 true ，协程就不会暂停，而是继续往下执行。
* await_suspend()，有一个类型为 std::coroutine_handle<> 的参数，返回类型可以是 void 或者 bool 。如果 await_ready() 返回了 false ，意味着协程要暂停，那么紧接着会调用这个函数。该函数的目的是用来接收协程句柄（也就是std::coroutine_handle<> 参数）并在异步操作完成的时候通过这个句柄让协程恢复执行。可以在这个函数上恢复协程的执行

* await_resume()，返回类型可以是 void ，也可以是其它类型，它的返回值就是 co_await 操作符的返回值。当协程恢复执行，或者不需要暂停的时候，会调用这个函数。

我们给出一个协程的用例：
```c++
class IntReader {
public:
    bool await_ready() {
        return false;
    }
    void await_suspend(std::coroutine_handle<> handle) {
        std::thread thread([this, handle]() {
            std::srand(static_cast<unsigned int>(std::time(nullptr)));
            value_ = std::rand();
            handle.resume();
        });
        thread.detach();
    }
    int await_resume() {
        return value_;
    }
private:
    int value_{};
};
class Task {
public:
    class promise_type {
    public:
        Task get_return_object() { return {}; }
        std::suspend_never initial_suspend() { return {}; }
        std::suspend_never final_suspend() noexcept { return {}; }
        void unhandled_exception() {}
        void return_void() {}
    };
};
Task PrintInt() {
    IntReader reader1;
    int total = co_await reader1;
    IntReader reader2;
    total += co_await reader2;
    IntReader reader3;
    total += co_await reader3;
    std::cout << total << std::endl;
}
```

在上述代码中，对于 await_ready() 函数，我们总是返回 false，即协程总是要暂停。
然后我们把子线程改成在 await_supsend() 中启动，也就是在协程暂停的时候来启动
在线程结束之后，将获取到的结果返回给调用者

此时执行流程是:
在 main() 函数中，主线程调用了 PrintInt() ，执行到 co_await reader1 这一行，协程暂停了，于是主线程退出 PrintInt() ，返回到 main() 继续往下执行，
最后在 while 循环中等待用户输入。
接下来，在 reader1中启动的子线程调用了协程句柄的 resume() ，所以从 co_await reader1 中恢复执行的是这个子线程，直到 co_await reader2 ，
协程再次暂停，子线程退出。

## 返回类型 和 promise_type
C++对协程的返回类型只有一个要求：包含名为 promise_type 的内嵌类型。
promise_type 需要符合C++规定的协议规范，也就是要定义几个特定的函数。 promise_type 是协程的一部分，当协程被调用，在堆上为其状态分配空间的时候，同时也会在其中创建一个对应的 promise_type 对象。通过在它上面定义的函数，我们可以与协程进行数据交互，以及控制协程的行为。

promise_type 要实现的第一个函数是 get_return_object() ，用来创建协程的返回值。在协程内，我们不需要显式地创建返回值，这是由编译器隐式调用 get_return_object() 来创建并返回的。这个关系看起来比较怪异， promise_type 是返回类型的内嵌类型，但编译器不会直接创建返回值，而是先创建一个 promise_type 对象，再通过这个对象来创建返回值。
<img src="https://zplutor.github.io/2022/03/25/cpp-coroutine-beginner/3.png">

### co_reture
在上述的例子中，使用co_reture捕获协程的返回数据
```c++
class Task {
public:
    class promise_type {
    public:
        promise_type() : value_(std::make_shared<int>()) {

        }
        Task get_return_object() { 
            return Task{ value_ };
        }
        void return_value(int value) {
            *value_ = value;
        }
        std::suspend_never initial_suspend() { return {}; }
        std::suspend_never final_suspend() noexcept { return {}; }
        void unhandled_exception() {}
    private:
        std::shared_ptr<int> value_;
    };
public:
    Task(const std::shared_ptr<int>& value) : value_(value) {

    }

    int GetValue() const {
        return *value_;
    }

private:
    std::shared_ptr<int> value_;
};
Task GetInt() {

    IntReader reader1;
    int total = co_await reader1;

    IntReader reader2;
    total += co_await reader2;

    IntReader reader3;
    total += co_await reader3;

    co_return total;
}
int main() {
    auto task = GetInt();
    std::string line;
    while (std::cin >> line) {
        std::cout << task.GetValue() << std::endl;
    }
    return 0;
}
```
co_return total 这个表达式等价于 promise_type.return_value(total) ，也就是说，返回的数据会通过 return_value() 函数传递给 promise_type 对象， promise_type 要实现这个函数才能接收到数据。除此之外，还要想办法让返回值 Task 能访问到这个数据。为了减少数据传递，我们可以在 promise_type 和 Task 之间共享同一份数据。

异步是具有传染性的， GetInt() 内部调用了异步操作，所以它自身实际上也是一个异步操作。为了等待它执行完成，我们把 task.GetValue() 的调用放在了 while 循环中，每当用户输入一次之后就进行输出。由于这是一个简单的示例程序，没有各种高级的异步同步机制，所以通过等待用户输入方式来变相地等待协程执行完成。

在真实的使用场景中，协程的返回类型还需要提供各种同步机制才能给调用者使用，例如加上回调、通知等，就像普通的异步操作一样。由此可见，协程的优点体现在它内部的代码逻辑上，而不是对外的使用方式上。当然，我们也可以让协程的返回类型实现 Awaitable 规范，让它可以被另外一个协程更好地调用。这样一来，调用协程的也必须是协程，这样层层往上传递，直到遇到不能改成协程的函数为止，例如 main() 函数。从这个角度来说，协程也是具有传染性的。

这里使用智能指针传递结果，是为了防止在协程释放内存之后，出现获取野指针的情况
# co_yield
当协程调用了 co_return ，意味着协程结束了，就跟我们在普通函数中用 return 结束函数一样。这时候，与这个协程实例有关的内存都会被释放掉，它不能再执行了。如果需要在协程中多次返回数据而不结束协程的话，可以使用 co_yield 操作符。

co_yield 的作用是，返回一个数据，并且让协程暂停，然后等下一次机会恢复执行。 co_yield value 这个表达式等价于 co_await promise_type.yield_value(value) ， co_yield 的参数会传递给 promise_type 的 yield_value() 函数，再把这个函数的返回值传给 co_await 。上文提到，传给 co_await 的参数要符合 Awaitable 规范，所以 yield_value() 的返回类型也要满足这个规范。在这里就可以使用预定义的 std::supsend_never 或 std::suspend_always ，通常会使用后者来让协程每次调用 co_yield 的时候都暂停。
例如：
```c++
#include <coroutine>
#include <iostream>
#include <thread>

class IntReader {
public:
    bool await_ready() {
        return false;
    }
    void await_suspend(std::coroutine_handle<> handle) {
        std::thread thread([this, handle]() {
            static int seed = 0;
            value_ = ++seed;
            handle.resume();
        });
        thread.detach();
    }
    int await_resume() {
        return value_;
    }
private:
    int value_{};
};

class Task {
public:
    class promise_type {
    public:
        Task get_return_object() { 
            return Task{ std::coroutine_handle<promise_type>::from_promise(*this) };
        }
        std::suspend_always yield_value(int value) {
            value_ = value;
            return {};
        }
        void return_void() { }
        std::suspend_never initial_suspend() { return {}; }
        std::suspend_never final_suspend() noexcept { return {}; }
        void unhandled_exception() {}

        int GetValue() const {
            return value_;
        }

    private:
        int value_{};
    };

public:
    Task(std::coroutine_handle<promise_type> handle) : coroutine_handle_(handle) {

    }
    int GetValue() const {
        return coroutine_handle_.promise().GetValue();
    }
    void Next() {
        coroutine_handle_.resume();
    }
private:
    std::coroutine_handle<promise_type> coroutine_handle_;
};

Task GetInt() {
    while (true) {
        IntReader reader;
        int value = co_await reader;
        co_yield value;
    }
}
int main() {
    auto task = GetInt();
    std::string line;
    while (std::cin >> line) {
        std::cout << task.GetValue() << std::endl;
        task.Next();
    }
    return 0;
}
```
为了让使用者可以恢复协程执行， Task 增加了一个 Next() 函数，这个函数调用了作为成员变量的协程句柄来恢复执行：
这意味着 Task 需要拿到协程的句柄，这是在 promise_type 的 get_return_object() 中通过以下方式传递过去的：

std::coroutine_handle 的 from_promise() 函数可以通过 promise_type 对象获取与之关联的协程句柄，反之，协程句柄上也有一个 promise() 函数可以获取对应的 promise_type 对象，他们是可以互相转换的。所以，在 Task 和 promise_type 之间就不需要使用 std::shared_ptr<int> 来共享数据了，Task通过协程句柄就能访问到 promise_type 对象。

# 协程的生命周期
C++会在堆上为协程的状态分配内存，这块内存必须在适当的时机来释放，否则就会造成内存泄漏。释放协程的内存有两种方式：自动释放和手动释放。
当协程结束的时候，如果我们不做任何干预，那么协程的内存就会被自动释放。调用了 co_return 语句之后，协程就会结束，下面两个协程是自动释放的例子：
```c++
Task GetInt() {
    IntReader reader;
    int value = co_await reader;
    co_return value;
}
Task PrintInt() {

    IntReader reader1;
    int value = co_await reader;
    std::cout << value << std::endl;
}
```
PrintInt() 没有出现 co_return 语句，编译器会在末尾隐式地加上 co_return 。

但是也存在问题，例如上述程序中，如果修改成：
```c++
Task GetInt() {
    co_return 1024;
}
int main() {
    auto task = GetInt();
    std::string line;
    while (std::cin >> line) {
        std::cout << task.GetValue() << std::endl;
    }
    return 0;
}
```
此时随着不断的随机访问进行，并不会一直访问1024，这是因为co_return之后协程被释放，再次GetValue只会捕获到野指针（这也是为什么前文中使用shared_ptr管理）
解决这个问题的方法是修改promise_type中的final_supsend()的返回类型，从 std::suspend_never 改成 std::suspend_always 。协程在结束的时候，会调用 final_suspend() 来决定是否暂停，如果这个函数返回了要暂停，那么协程不会自动释放，此时协程句柄还是有效的，可以安全访问它内部的数据。

但是此时就需要手动释放Task了
```c++
~Task() {
    coroutine_handle_.destroy();
}
```

与 final_suspend() 相对应的是 initial_suspend() ，在协程刚开始执行的时候，会调用这个函数来决定是否暂停。我们可以将这个函数的返回类型改成 std::suspend_always 来让协程一执行即暂停。这对于一些需要延迟执行的场景是有用的，例如，我们想先获取一批协程句柄，像数据那样对它们进行管理，在稍后的时机再挑选合适的协程来执行。
```c++
std::suspend_never initial_suspend() { return {}; }
std::suspend_never final_suspend() noexcept { return {}; }
```


# 参考链接:
https://zplutor.github.io/2022/03/25/cpp-coroutine-beginner/